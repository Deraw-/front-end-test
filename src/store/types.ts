// tslint:disable-next-line:no-empty-interface
export interface RootState {}

export interface Product extends Object {
	id: number;
	quantity: number;
}

export interface CartProduct extends Product {
	name: string;
	price: number;
	image: string;
}

// This is used to properly type components using mappers from Vuex
// see https://github.com/vuejs/vuex/pull/1121
// tslint:disable-next-line: no-empty-interface
export default interface VuexBindings {
	cartProducts: CartProduct[];
}
