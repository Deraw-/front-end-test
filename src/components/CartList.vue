<template>
	<ul class="cart-list">
		<li
			v-for="product in products"
			:key="product.id"
		>
			<div class="product-infos">
				<img
					:src="`/api/img/${product.image}`"
					alt=""
					class="image"
				>

				<div class="name">
					{{ product.name }}

					<button
						class="btn primary"
						@click="deleteFromCart(product)"
					>
						Delete
					</button>
				</div>
			</div>

			<UpdateCartBtn
				:product="product"
				class="cart-btn"
			/>

			<div class="price">
				{{ product.price | euros }}
			</div>
		</li>
	</ul>
</template>

<script lang="ts">
	import Vue from 'vue';
	import { mapActions } from 'vuex';

	import UpdateCartBtn from '@/components/UpdateCartBtn.vue';

	export default Vue.extend({
		name: 'CartList',
		components: {
			UpdateCartBtn
		},
		props: {
			products: {
				type: [Object, Array],
				required: true
			}
		},
		methods: {
			...mapActions('cart', ['deleteFromCart'])
		}
	});
</script>

<style lang="scss" scoped>
	.cart-list {
		margin-top: 25px;

		li {
			display: flex;
			flex-wrap: wrap;
			align-items: center;

			padding: 15px 20px;

			border-top: 2px solid #f5f5f5;

			&:nth-child(even) {
				background: #f5f5f5;
			}

			> div {
				width: calc(100% / 3);
			}
		}
	}

	.product-infos {
		display: flex;
		flex-wrap: wrap;
	}

	.image {
		width: 50px;
		height: 50px;
		margin-right: 15px;

		border-radius: 50%;
		object-fit: cover;
	}

	.name {
		display: flex;
		flex-direction: column;
		align-items: flex-start;
	}

	.cart-btn {
		text-align: center;
	}

	.price {
		font-weight: 700;
		color: #008393;

		text-align: right;
	}

	.btn {
		margin: 0;
		padding: 0;
		margin-top: 5px;

		background: transparent;

		&:hover,
		&:focus {
			color: #333;
			background: transparent;
		}
	}

	@media screen and (max-width: 700px) {
		.cart-list li {
			display: block;

			> div {
				text-align: left;

				+ div {
					margin-top: 15px;
				}
			}
		}
	}
</style>
