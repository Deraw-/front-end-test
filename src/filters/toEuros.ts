const toEuros = (value: number): string => {
	const price = `${value.toFixed(2)} â‚¬`;

	return price.replace('.', ',');
};

export default toEuros;
