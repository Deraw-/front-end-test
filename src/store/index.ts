import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import VuexPersistence from 'vuex-persist';
import { RootState } from './types';

// Modules
import cart from './modules/cart';
import products from './modules/products';

// The store is saved in the browser's session
const vuexLocal = new VuexPersistence({
	storage: window.sessionStorage
});

Vue.use(Vuex);

// See https://vuex.vuejs.org/fr/getting-started.html for help
const store: StoreOptions<RootState> = {
	strict: true,
	state: {},
	// See https://vuex.vuejs.org/guide/modules.html for more info on modules
	modules: {
		cart,
		products
	},
	plugins: [
		vuexLocal.plugin
	]
};

export default new Vuex.Store<RootState>(store);
