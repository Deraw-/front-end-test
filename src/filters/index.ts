import Vue from 'vue';

import toEuros from './toEuros';

Vue.filter('euros', toEuros);
