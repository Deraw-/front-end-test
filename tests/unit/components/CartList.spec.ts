import { shallowMount } from '@vue/test-utils';
import CartList from '@/components/CartList.vue';

import localVue from 'tests-unit/index';
import store from '@/store/';
import '@/filters/';

// Data
import { CartProduct } from '@/store/types';
const products: CartProduct[] = [];

for (let index = 0; index < 8; index++) {
	products.push({
		id: index,
		quantity: 15,
		name: 'name',
		price: 15,
		image: '/img.jpg'
	});
}

// Build component before each test
let wrapper: any;
beforeEach(() => {
	wrapper = shallowMount(CartList, {
		localVue,
		store,
		propsData: {
			products
		}
	});
});

// Tests
describe('CartList.vue', () => {
	it('renders correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
