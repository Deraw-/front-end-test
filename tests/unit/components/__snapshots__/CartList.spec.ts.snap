// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartList.vue renders correctly 1`] = `
<ul class="cart-list">
  <li>
    <div class="product-infos"><img src="/api/img//img.jpg" alt="" class="image">
      <div class="name">
        name

        <button class="btn primary">
          components.cartList.deleteBtn
        </button></div>
    </div>
    <updatecartbtn-stub product="[object Object]" class="cart-btn"></updatecartbtn-stub>
    <div class="price">
      15,00 €
    </div>
  </li>
  <li>
    <div class="product-infos"><img src="/api/img//img.jpg" alt="" class="image">
      <div class="name">
        name

        <button class="btn primary">
          components.cartList.deleteBtn
        </button></div>
    </div>
    <updatecartbtn-stub product="[object Object]" class="cart-btn"></updatecartbtn-stub>
    <div class="price">
      15,00 €
    </div>
  </li>
  <li>
    <div class="product-infos"><img src="/api/img//img.jpg" alt="" class="image">
      <div class="name">
        name

        <button class="btn primary">
          components.cartList.deleteBtn
        </button></div>
    </div>
    <updatecartbtn-stub product="[object Object]" class="cart-btn"></updatecartbtn-stub>
    <div class="price">
      15,00 €
    </div>
  </li>
  <li>
    <div class="product-infos"><img src="/api/img//img.jpg" alt="" class="image">
      <div class="name">
        name

        <button class="btn primary">
          components.cartList.deleteBtn
        </button></div>
    </div>
    <updatecartbtn-stub product="[object Object]" class="cart-btn"></updatecartbtn-stub>
    <div class="price">
      15,00 €
    </div>
  </li>
  <li>
    <div class="product-infos"><img src="/api/img//img.jpg" alt="" class="image">
      <div class="name">
        name

        <button class="btn primary">
          components.cartList.deleteBtn
        </button></div>
    </div>
    <updatecartbtn-stub product="[object Object]" class="cart-btn"></updatecartbtn-stub>
    <div class="price">
      15,00 €
    </div>
  </li>
  <li>
    <div class="product-infos"><img src="/api/img//img.jpg" alt="" class="image">
      <div class="name">
        name

        <button class="btn primary">
          components.cartList.deleteBtn
        </button></div>
    </div>
    <updatecartbtn-stub product="[object Object]" class="cart-btn"></updatecartbtn-stub>
    <div class="price">
      15,00 €
    </div>
  </li>
  <li>
    <div class="product-infos"><img src="/api/img//img.jpg" alt="" class="image">
      <div class="name">
        name

        <button class="btn primary">
          components.cartList.deleteBtn
        </button></div>
    </div>
    <updatecartbtn-stub product="[object Object]" class="cart-btn"></updatecartbtn-stub>
    <div class="price">
      15,00 €
    </div>
  </li>
  <li>
    <div class="product-infos"><img src="/api/img//img.jpg" alt="" class="image">
      <div class="name">
        name

        <button class="btn primary">
          components.cartList.deleteBtn
        </button></div>
    </div>
    <updatecartbtn-stub product="[object Object]" class="cart-btn"></updatecartbtn-stub>
    <div class="price">
      15,00 €
    </div>
  </li>
</ul>
`;
